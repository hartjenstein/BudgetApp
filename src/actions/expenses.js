import uuid from 'uuid';
import database from '../firebase/firebase';

// Action generators for expenses reducer //

///* flow without firebase:
// component calls actoin generator
// action generator returns object
// component dispatches object
// redux store changes

///*flow with firebase
// component calls action generator
// action generator returns function
// component dispatches function (?)
// function runs (has the ability to dispatch other actions and do whatever it wants)
//redux by default does not allow to dispatch functions. Thats what we use thunk for

//ADD_EXPENSE
//action generator without firebase
/* export const addExpense = ({ 
    description = '', 
    note = '', 
    amount = 0, 
    createdAt = 0 
} = {}) =>  ({
    type: 'ADD_EXPENSE',
    expense: {
        id: uuid(),
        description,
        note,
        amount,
        createdAt
}
}); */
// the ADD_EXPENSE action generator (above) is simplefied and we add the  async action gen. startAddExpense, as firebase database only takes objects and 
// the redux store state only accepts arrays.
export const addExpense = (expense) => ({
	type: 'ADD_EXPENSE',
	expense
});

export const startAddExpense = (expenseData = {}) => {
  // passing dispatch as an arguement, this is happening internally with thunk middleware and redux! 
	return (dispatch) => {
		const {
			//moving defaults here from ADD_EXPENSE, destructuring inside function instead of in the arguements to make it readable more easily
			description = '',
				note = '',
				amount = 0,
				createdAt = 0
		} = expenseData;
		const expense = { description, note, amount, createdAt };
		// saving data to firebase, then dispatching the action
		// asyncronus action generator - promise comes back from firebase 
		//and gets called with ref as an arguement - we get the ref object returned from our call to the database
		// ref.key contains the id firebase created for data set we've just pushed
		// 
		return database.ref('expenses')
			.push(expense)
			.then((ref) => {
				dispatch(addExpense({
					id: ref.key,
					...expense
				}));
		});
	};
};

// REMOVE_EXPENSE 
export const removeExpense = ({ id } = {}) => ({
	type: 'REMOVE_EXPENSE',
	id
});

// EDIT_EXPENSE 
export const editExpense = (id, updates) => ({
	type: 'EDIT_EXPENSE',
	id,
	updates
});

export const startEditExpense = (id, updates) => {
  return (dispatch) => {
    return database.ref(`expenses/${id}`).update({
      ...updates
    }).then(() => {
      dispatch(editExpense(id, updates))
    });
  };
}; 

// SET_EXPENSES
export const setExpenses = (expenses) => ({
  type: 'SET_EXPENSES',
  expenses
});

//Async function to handle firebase set data
export const startSetExpenses = () => {
  // dispatch arguement is provided internally from thunk middleware 
 return (dispatch) => {
    return database.ref('expenses').once('value').then((snapshot) => {
      const expenses = [];
      snapshot.forEach((childSnapshot) => {
        expenses.push({
          id: childSnapshot.key,
          ...childSnapshot.val()
        });
      });
      dispatch(setExpenses(expenses));
    });
  }
};

export const startRemoveExpense = ({ id } = {}) => {
  return (dispatch) => { // dispatch arguement gets passed in by redux library / redux thunk middleware
    return database.ref(`expenses/${id}`).remove().then(() => {
      dispatch(removeExpense({ id }));
    });
  };
};