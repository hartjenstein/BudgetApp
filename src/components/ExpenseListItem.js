import React from 'react';
import { Link } from 'react-router-dom';
import moment from 'moment';
 
const ExpenseListItem = ({ id, description, amount, createdAt, note  } ) => (
    <div className="ExpenseListItem">
        {description &&  <Link to={`/edit/${id}`}><h3>{description}</h3></Link>}
        {note &&  <p>Note: {note}</p>}
        {amount &&  <p>Amount: {amount}</p>}
        {createdAt &&  <p>Time created: {moment(createdAt).format('MM Do, YYYY')}</p>}
    </div>
)

export default ExpenseListItem;