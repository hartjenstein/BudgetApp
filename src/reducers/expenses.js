// Expenses Reducer

const expensesReducerDefaultState = [];
export default (state = expensesReducerDefaultState, action) => {
    switch (action.type) {
        case 'ADD_EXPENSE': 
        //use concat instead of push to avoid mutating the array
        // here we use array destructuring and the rest operator
          return [...state, action.expense];
        case 'REMOVE_EXPENSE': 
            return state.filter(({ id }) => id !== action.id);
        case 'EDIT_EXPENSE': 
            return state.map((expense) => {
                if (expense.id === action.id) {
                    return {...expense,
                        ...action.updates
                    };
                } else {
                    return expense
                }
            });
        case 'SET_EXPENSES': 
            return action.expenses;
        default: 
            return state;
    }
};
