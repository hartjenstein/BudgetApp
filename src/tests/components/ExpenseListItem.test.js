import React from 'react';
import { shallow } from 'enzyme';
import ExpenseListItem  from '../../components/ExpenseListItem';
import expenses from '../fixtures/expenses';
const expense = expenses[0]
test('should render ExpenseList Item', () => {
    const wrapper = shallow(<ExpenseListItem {...expenses[0]}/>);
    expect(wrapper).toMatchSnapshot();
});
