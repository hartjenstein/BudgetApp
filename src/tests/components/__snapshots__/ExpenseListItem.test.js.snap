// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render ExpenseList Item 1`] = `
<div
  className="ExpenseListItem"
>
  <Link
    replace={false}
    to="/edit/1"
  >
    <h3>
      gum
    </h3>
  </Link>
  <p>
    Amount: 
    $0.12
  </p>
  0
</div>
`;
